偶发SimpleDraweeView初始化时崩溃

android.view.InflateException: Binary XML file line #26 in com.miui.hybrid:layout/titlebar_dialog_view: Binary XML file line #26 in com.miui.hybrid:layout/titlebar_dialog_view: Error inflating class com.facebook.drawee.view.SimpleDraweeView\nCaused by: android.view.InflateException: Binary XML file line #26 in com.miui.hybrid:layout/titlebar_dialog_view:

Error inflating class com.facebook.drawee.view.SimpleDraweeView\nCaused by: java.lang.reflect.InvocationTargetException
at java.lang.reflect.Constructor.newInstance0(Native Method)
at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
at android.view.LayoutInflater.createView(LayoutInflater.java:858)
at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:1014)
at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:965)
at android.view.LayoutInflater.rInflate(LayoutInflater.java:1144)
at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1105)
at android.view.LayoutInflater.rInflate(LayoutInflater.java:1147)
at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1105)
at android.view.LayoutInflater.rInflate(LayoutInflater.java:1147)
at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1105)
at android.view.LayoutInflater.inflate(LayoutInflater.java:686)
at android.view.LayoutInflater.inflate(LayoutInflater.java:538)
at android.view.LayoutInflater.inflate(LayoutInflater.java:481)
at com.android.internal.policy.PhoneWindow.setContentView(PhoneWindow.java:442)

at android.app.Dialog.setContentView(Dialog.java:573)
at org.hapjs.component.view.BaseTitleDialog.initView(SourceFile:93)
at org.hapjs.component.view.BaseTitleDialog.<init>(SourceFile:79)
at org.hapjs.component.view.MenubarView.initDialog(SourceFile:467)
at org.hapjs.component.view.MenubarView.initTitlebarView(SourceFile:139)
at org.hapjs.component.view.MenubarView.<init>(SourceFile:107)
at org.hapjs.render.Display.ensureMenuBarView(SourceFile:480)
at org.hapjs.render.Display.setupMenuBar(SourceFile:1269)
at org.hapjs.render.Display.setup(SourceFile:198)
at org.hapjs.render.DecorLayout.setupDisplay(SourceFile:101)
at org.hapjs.render.vdom.DocComponent.attachChildren(SourceFile:110)
at org.hapjs.render.vdom.VDocument.attachChildren(SourceFile:35)
at org.hapjs.render.RootView.forward(SourceFile:1436)
at org.hapjs.render.RootView.onPageChangedInMainThread(SourceFile:1385)
at org.hapjs.render.RootView.onPageChanged(SourceFile:1312)
at org.hapjs.render.PageManager.push(SourceFile:369)
at org.hapjs.render.PageManager$HandlerImpl.handleMessage(SourceFile:542)

at android.os.Handler.dispatchMessage(Handler.java:107)
at android.os.Looper.loop(Looper.java:225)
at android.app.ActivityThread.main(ActivityThread.java:7566)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:539)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:994)\nCaused by: java.lang.NullPointerException: SimpleDraweeView was not initialized!
at com.facebook.common.internal.Preconditions.checkNotNull(SourceFile:231)
at com.facebook.drawee.view.SimpleDraweeView.init(SourceFile:87)
at com.facebook.drawee.view.SimpleDraweeView.<init>(SourceFile:64)
... 38 more

 

原因：

Runtime中FrescoUtils是异步初始化的，需要保证使用前完成初始化

--- a/core/runtime/android/runtime/src/main/java/org/hapjs/component/view/BaseTitleDialog.java
+++ b/core/runtime/android/runtime/src/main/java/org/hapjs/component/view/BaseTitleDialog.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, the hapjs-platform Project Contributors
+ * Copyright (c) 2021-2022, the hapjs-platform Project Contributors
  * SPDX-License-Identifier: Apache-2.0
  */
 
@@ -37,6 +37,7 @@ import java.util.List;
 import org.hapjs.bridge.HybridManager;
 import org.hapjs.bridge.HybridView;
 import org.hapjs.bridge.LifecycleListener;
+import org.hapjs.common.utils.FrescoUtils;
 import org.hapjs.model.MenubarItemData;
 import org.hapjs.render.Display;
 import org.hapjs.runtime.ConfigurationManager;
@@ -76,22 +77,23 @@ public class BaseTitleDialog extends Dialog implements ConfigurationManager.Conf
 
     public BaseTitleDialog(@NonNull Context context) {
         super(context, R.style.HapTheme_Dialog);
-        initView();
+        initView(context);
     }
 
     public BaseTitleDialog(@NonNull Context context, int themeResId) {
         super(context, themeResId);
-        initView();
+        initView(context);
     }
 
     protected BaseTitleDialog(
             @NonNull Context context, boolean cancelable,
             @Nullable OnCancelListener cancelListener) {
         super(context, cancelable, cancelListener);
-        initView();
+        initView(context);
     }
 
-    private void initView() {
+    private void initView(Context context) {
+        FrescoUtils.initialize(context.getApplicationContext());
         setContentView(R.layout.titlebar_dialog_view);
         mLocationStr = getContext().getResources().getString(R.string.menubar_dlg_using_location);
         mRecordingStr = getContext().getResources().getString(R.string.menubar_dlg_using_record);

